Fix wrong PictureID rolling over in VP9 and VP8

jcague opened the pull request:
VP9 and VP8 payloads include the PictureID. This field has a length of 15 bits, but VP8.cpp and VP9.cpp were using a SeqManager with uint16_t, which is 16 bits. 

This may cause different issues. I've reproduced very low framerates in VP9/SVC, but there could also be video artifacts, frozen videos, etc. VP8 also translates PictureIDs so that it could have an additional impact. The problem I saw is that we were marking all VP9/VP8 packets as `isOldPacket` after rolling over the PictureID. So we started to send higher temporal and spatial layers, and it broke VP9/SVC decoding.

I'm adding a bit length parameter when creating the SeqManager (`SeqManager<uint16_t, 15>`) to handle PictureID properly. I've added some tests for this case too.

jcague committed:
Fix wrong PictureID rollover in VP9 and VP8

ibc commented:
Amazing. Let's review this week ASAP.

jmillan commented on worker/test/src/RTC/Codecs/TestVP9.cpp:
```diff
@@ -0,0 +1,76 @@
+		context.SetTargetSpatialLayer(0);
+
+		// Frame 1
+		auto forwarded = ProcessVP9Packet(context, 32767, 0);
```
just for readability, can you use this representation for `pictureId`?

```c++
auto pictureId = (1 << 15) - 1;
```

Same for VP9 test.

jmillan commented on worker/include/RTC/SeqManager.hpp:
```diff
@@ -7,11 +7,11 @@
 namespace RTC
 {
-	template<typename T>
+	template<typename T, unsigned int N = 0>
```
Using `uint8_t` type looks more concise IMHO and it's more than enough space to hold all the possible values.

```suggestion
	template<typename T, uint8_t N = 0>
```

jmillan replied:
Also please add a little comment indicating that `N` represents the number of bits to be used in `T`.

jmillan commented on worker/src/RTC/SeqManager.cpp:
```diff
@@ -121,21 +121,22 @@ namespace RTC
 	// Explicit instantiation to have all SeqManager definitions in this file.
 	template class SeqManager<uint8_t>;
 	template class SeqManager<uint16_t>;
+	template class SeqManager<uint16_t, 15>; // For PictureID (15bits).
```
```suggestion
	template class SeqManager<uint16_t, 15>; // For PictureID (15 bits).
```

jmillan reviewed (commented):
Very nice!

Could you please add a test for SeqManager too?

ibc commented on worker/src/RTC/SvcConsumer.cpp:
```diff
@@ -586,6 +586,7 @@ namespace RTC
 				MS_DEBUG_TAG(rtp, "sync key frame received");
 
 			this->rtpSeqManager.Sync(packet->GetSequenceNumber() - 1);
+			this->encodingContext->SyncRequired();
```
Why is this added? It doesn't seem really related to changes in this PR. Is it a missing call?

@jmillan can you confirm please?

jmillan replied:
I believe this is OK, so we don't re-use any previous PictureId and completely start from a new one.

Is that the intention @jcague?

jcague replied:
It's not exactly like that, but you are close. SvcConsumer uses the SeqManager only to know if a PictureID is old (it doesn't translate it as SimulcastConsumer does). The isOldPacket gets calculated with the MaxInput value inside the SeqManager. And it looks like the right way to initialize MaxInpu is by calling SeqManager::Sync().

This is what I inferred from the code yesterday. Do you think it makes sense?

jmillan replied:
It makes sense to me

ibc commented on worker/test/src/RTC/Codecs/TestVP8.cpp:
```diff
@@ -292,4 +292,32 @@ SCENARIO("process VP8 payload descriptor", "[codecs][vp8]")
+		context.SetCurrentTemporalLayer(0);
+		context.SetTargetTemporalLayer(0);
+
+		// Frame 1
```
```suggestion
		// Frame 1.
```

jcague replied:
Should I update the same comments in TestVP9?

ibc replied:
Yes please, dot at the end in all comments. ☺️

ibc commented on worker/test/src/RTC/Codecs/TestVP8.cpp:
```diff
@@ -292,4 +292,32 @@ SCENARIO("process VP8 payload descriptor", "[codecs][vp8]")
+		REQUIRE(forwarded->pictureId == 1);
+		REQUIRE(forwarded->tl0PictureIndex == 1);
+
+		// Frame 2
```
```suggestion
		// Frame 2.
```

ibc commented on worker/test/src/RTC/Codecs/TestVP8.cpp:
```diff
@@ -292,4 +292,32 @@ SCENARIO("process VP8 payload descriptor", "[codecs][vp8]")
+		REQUIRE(forwarded->pictureId == 2);
+		REQUIRE(forwarded->tl0PictureIndex == 1);
+
+		// Frame 3
```
```suggestion
		// Frame 3.
```

ibc commented on worker/test/src/RTC/Codecs/TestVP9.cpp:
```diff
@@ -0,0 +1,76 @@
+		context.SetCurrentSpatialLayer(0);
+		context.SetTargetSpatialLayer(0);
+
+		// Frame 1
```
```suggestion
		// Frame 1.
```

ibc commented on worker/test/src/RTC/Codecs/TestVP9.cpp:
```diff
@@ -0,0 +1,76 @@
+		REQUIRE(forwarded);
+		REQUIRE(forwarded->pictureId == 32767);
+
+		// Frame 2
```
```suggestion
		// Frame 2.
```

ibc commented on worker/test/src/RTC/Codecs/TestVP9.cpp:
```diff
@@ -0,0 +1,76 @@
+		REQUIRE(forwarded);
+		REQUIRE(forwarded->pictureId == 0);
+
+		// Frame 3
```
```suggestion
		// Frame 3.
```

jcague committed:
Fix wrong delta calculation inside SeqManager

jcague commented:
I've updated how we calculate deltas inside SeqManager to fix an issue I found executing long-running tests locally. And added more unit tests to cover that case properly.

jcague committed:
Fix lint issues

ibc commented:
Just a question: somewhere we shift the payload or VP8 and VP9 (if the original received packet uses 8 bits for pictureId and the new computed pictureId value in the Consumer doesn't fit into 8 bits):
- Here we should shift the field if the new value doesn't fit into 7 bits and convert it into 16 bits and set first one to 1 to indicate that it's a 16 bits field, am I right?
- Or should we respect the size of the pictureId field in the original packet and just use 7 bits?

jmillan commented on worker/test/src/RTC/Codecs/TestVP8.cpp:
```diff
@@ -5,6 +5,8 @@
 
 using namespace RTC;
 
+constexpr uint16_t kMaxPictureId = (1 << 15) - 1;
```
why the `k` prefix? :-). We don't use it in the codebase.

ibc replied:
Please replace with `MaxPictureId`.

jcague replied:
My fault. I did a quick search and found some of them. But they were inside the libwebrtc folder.

jcague committed:
Apply comments in the PR

jcague committed:
Apply comments in the PR

jcague commented:
SimulcastConsumer shifts the payload in VP8 because there is a translation. But SvcConsumer doesn't translate pictureIds in VP9/SVC, so I think there's no need to shift the payload and force the two-byte header.

ibc commented:
I've been told by @lminiero that "the only problem with 7-bit PictureID is that browsers often freeze video when they're used, because of a known bug, but no one uses 7-bit anymore apparently, older versions of ffmpeg were the only ones".

So I'd like to ensure what exactly we are doing when it comes to deal with 7bit pictureID.

<img width="761" alt="CleanShot-2023-01-19-at-10 40 23@2x" src="https://user-images.githubusercontent.com/16191/213408253-6fa2d7a5-6d90-42dd-8c6a-ce10073ac7ad.png">

So in VP8, we always force pictureId to be 15-bits and we properly set the first bit to 1 (so it indicates that 15 bits are used for the value). So IMHO we are good.

ibc commented on worker/src/RTC/SeqManager.cpp:
```diff
@@ -7,43 +7,50 @@
+	template<typename T, uint8_t N>
+	T SeqManager<T, N>::Delta(const T lhs, const T rhs)
+	{
+		T value = (lhs > rhs) ? (lhs - rhs) : (MaxValue - rhs + lhs);
+		return value & MaxValue;
```
Let's please always add a blank line before `return`.
```suggestion
		T value = (lhs > rhs) ? (lhs - rhs) : (MaxValue - rhs + lhs);

		return value & MaxValue;
```

ibc committed:
Add blank line before return

ibc commented:
Please apply this diff in `CHANGELOG.md` file and IMHO we are done:

```diff
diff --git a/CHANGELOG.md b/CHANGELOG.md
index a632a5a4e..42076219a 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,11 @@
 # Changelog
 
 
+### Next
+
+* Fix wrong `PictureID` rolling over in VP9 and VP8 ([PR #984(https://github.com/versatica/mediasoup/pull/984) by @jcague).
+
+
 ### 3.11.5
 
 * Require Node.js >= 16 ([PR #973(https://github.com/versatica/mediasoup/pull/973)).
```

ibc reviewed (approved):
IMHO there is nothing missing in this PR and we are good. @jmillan can we merge?

lminiero commented:
> So in VP8, we always force pictureId to be 15-bits and we properly set the first bit to 1 (so it indicates that 15 bits are used for the value). So IMHO we are good.

That depends on what you're doing. If you're leaving the packet as it is, setting the M bit to 1 with a PictureID that is 7 bit, means the recipient will read the following byte as well thinking it's part of the PictureID, when it isn't. This will cause errors reading the payload too, as everything will be read shifted by a byte, which will cause ugly artifacts. The only way to really force 15-bit PictureID also when using 7-bit PictureID is to shift the whole payload after that point by 1 byte, so that you reserve room for a 15-bit PictureID.

ibc commented:
> That depends on what you're doing. If you're leaving the packet as it is, setting the M bit to 1 with a PictureID that is 7 bit, means the recipient will read the following byte as well thinking it's part of the PictureID, when it isn't. 

As per attached code screenshot above we don't do that. We always convert to 15-bits and set first bit to 1.

I mean, we never let a vp8 payload with 7-bits pictureId pass, we always convert it to 15-bits.

lminiero commented:
> As per attached code screenshot above we don't do that. We always convert to 15-bits and set first bit to 1.
> 
> I mean, we never let a vp8 payload with 7-bits pictureId pass, we always convert it to 15-bits.

Ack, apologies, I didn't look at the code :grin:

ibc commented:
> Ack, apologies, I didn't look at the code 😁

Ok, you didn't look at the text either 😂

> So in VP8, we always force pictureId to be 15-bits and we properly set the first bit to 1 (so it indicates that 15 bits are used for the value). So IMHO we are good.

ibc commented:
Merging this. Thanks a lot @jcague

ibc commented:
Released in 3.11.6.

fippo commented:
> but no one uses 7-bit anymore apparently

Not after @ggarber made libwebrtc emit the two-byte version unconditionally a couple of years back
